src/abstract-api-adapter.coffee

```diff
@@ -0,0 +1,103 @@
+request = require 'request'
+try
+	{ Adapter } = require 'hubot'
+catch
+	prequire = require 'parent-require'
+	{ Adapter } = prequire 'hubot'
+
+class AbstractAPIAdapter extends Adapter
+	constructor: ->
+		super
+		@interval = 1000
+		@lastReport = null
+		@allowedToPoll = true
+		throw new TypeError('Abstract class')
+
+	run: ->
+		setInterval @maybePoll, @interval
+		@emit 'connected'
+
+	maybePoll: =>
+		if @allowedToPoll
+			@allowedToPoll = false
+			@poll()
+
+	# Execute and callback a series of paged requests until we run out of pages or a filter rejects
+	getUntil: (options, each, filter, done) ->
+		request.get? options, (error, response, body) =>
+			if error or response?.statusCode isnt 200
+				if response?.statusCode in [429, 503]
+					@report { error: error, html: response?.statusCode, url: options.url }
+					# Retry if the error code indicates temporary outage
+					@getUntil options, filter, each, done
+				else
+					@robot.logger.error response?.statusCode + '-' + options.url
+					@processDone error, null, { error: error, html: response?.statusCode, url: options.url }, done
+			else
+				responseObject = JSON.parse(body)
+				results = @extractResults responseObject
+				for result in results
+					if (not filter?) or filter result
+						each result
+					else
+						@processDone null, results, { html: response?.statusCode, url: options.url }, done
+						return
+				if @extractNext responseObject
+					options.url = @extractNext responseObject
+					@getUntil options, filter, each, done
```

Comment:
Flipped filter/each here too.

This does suggest we haven't really tested this code... Can we have some automated tests for this somewhere to catch these issues? Here if that's easy, or an integration test in an example extending class somewhere else would be great.